lib/bb/runqueue.py

```diff
@@ -0,0 +1,3116 @@
+"""
+BitBake 'RunQueue' implementation
+
+Handles preparation and execution of a queue of tasks
+"""
+
+# Copyright (C) 2006-2007  Richard Purdie
+#
+# SPDX-License-Identifier: GPL-2.0-only
+#
+
+import copy
+import os
+import sys
+import stat
+import errno
+import logging
+import re
+import bb
+from bb import msg, event
+from bb import monitordisk
+import subprocess
+import pickle
+from multiprocessing import Process
+import shlex
+import pprint
+
+bblogger = logging.getLogger("BitBake")
+logger = logging.getLogger("BitBake.RunQueue")
+hashequiv_logger = logging.getLogger("BitBake.RunQueue.HashEquiv")
+
+__find_sha256__ = re.compile( r'(?i)(?<![a-z0-9])[a-f0-9]{64}(?![a-z0-9])' )
+
+def fn_from_tid(tid):
+     return tid.rsplit(":", 1)[0]
+
+def taskname_from_tid(tid):
+    return tid.rsplit(":", 1)[1]
+
+def mc_from_tid(tid):
+    if tid.startswith('mc:') and tid.count(':') >= 2:
+        return tid.split(':')[1]
+    return ""
+
+def split_tid(tid):
+    (mc, fn, taskname, _) = split_tid_mcfn(tid)
+    return (mc, fn, taskname)
+
+def split_mc(n):
+    if n.startswith("mc:") and n.count(':') >= 2:
+        _, mc, n = n.split(":", 2)
+        return (mc, n)
+    return ('', n)
+
+def split_tid_mcfn(tid):
+    if tid.startswith('mc:') and tid.count(':') >= 2:
+        elems = tid.split(':')
+        mc = elems[1]
+        fn = ":".join(elems[2:-1])
+        taskname = elems[-1]
+        mcfn = "mc:" + mc + ":" + fn
+    else:
+        tid = tid.rsplit(":", 1)
+        mc = ""
+        fn = tid[0]
+        taskname = tid[1]
+        mcfn = fn
+
+    return (mc, fn, taskname, mcfn)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears you are unpacking more than three variables while using a function that returns more than one variable. Using a large number of return values is prone to errors. We recommend that you return a small class or `namedtuple` instance instead.

[Learn more](https://docs.python.org/3/library/typing.html#typing.NamedTuple)